### DISCUSSION TITLE
Auditing with mikro-orm

### DISCUSSION THREAD

**Discussion by sebastianhaberey** [mikro-orm/mikro-orm#1621] [Q&A]
I'm looking for a way to implement auditing (entity history / change tracking) for my application. The ORM layer might be a fitting place for this. The concept of change sets is already there and the auditing isn't in any way specific to my application. Hibernate, for example, offers [Envers](https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#envers) for that purpose.

Now I'm not asking for this as a feature, I can see how this is a bit out of scope. My question is, would you advise me to implement this in my application layer, or is there another, smarter way where I could hook into mikro-orm or reuse stuff? How would you go about doing something like this? Any suggestions?

**Comment by B4nan** [comment]
Check the `onFlush` event, it allows to hook into the change set tracking and do pretty much whatever you want, e.g. you can handle soft deletes this way:

https://github.com/mikro-orm/mikro-orm/issues/1492#issuecomment-785394397

https://mikro-orm.io/docs/lifecycle-hooks/#using-onflush-event
